components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java

```diff
@@ -782,4 +794,143 @@ protected boolean isNoPassive(Response response) {
+
+        int timeStampSkewInSeconds = ssoAgentConfig.getSAML2().getTimeStampSkewInSeconds();
+
+        DateTime validFrom = assertion.getConditions().getNotBefore();
+        DateTime validTill = assertion.getConditions().getNotOnOrAfter();
+
+        if (validFrom != null && validFrom.minusSeconds(timeStampSkewInSeconds).isAfterNow()) {
+            throw new SSOAgentException("Failed to meet SAML Assertion Condition 'Not Before'");
+        }
+
+        if (validTill != null && validTill.plusSeconds(timeStampSkewInSeconds).isBeforeNow()) {
+            throw new SSOAgentException("Failed to meet SAML Assertion Condition 'Not On Or After'");
+        }
+
+        if (validFrom != null && validTill != null && validFrom.isAfter(validTill)) {
+            throw new SSOAgentException("SAML Assertion Condition 'Not Before' must be less than the value of 'Not On" +
+                    " " +
+
+                    "Or After'");
+        }
+    }
+
+    /**
+     * Validates the XML Signature object
+     *
+     * @param signature XMLObject
+     * @throws SSOAgentException
+     */
+
+    private void validateSignature(XMLObject signature) throws SSOAgentException{
+
+        SignatureImpl signImpl = (SignatureImpl) signature;
+        try {
+            SAMLSignatureProfileValidator signatureProfileValidator = new SAMLSignatureProfileValidator();
+            signatureProfileValidator.validate(signImpl);
+
+            // Following code segment is taken from org.opensaml.security.SAMLSignatureProfileValidator
+            // of OpenSAML 2.6.4. This is done to get the latest XSW related fixes.
+            XMLSignature apacheSig = signImpl.getXMLSignature();
+            SignableSAMLObject signableObject = (SignableSAMLObject) signature.getParent();
+
+            Reference ref = null;
+            try {
+                ref = apacheSig.getSignedInfo().item(0);
+            } catch (XMLSecurityException e) {
+                // This exception should never occur, because it's already checked
+                // from the previous call to signatureProfileValidator#validate
+                log.error("Apache XML Security exception obtaining Reference", e);
+                throw new ValidationException("Could not obtain Reference from Signature/SignedInfo", e);
+            }
+
+            String uri = ref.getURI();
+
+            validateReferenceURI(uri, signableObject);
+            validateObjectChildren(apacheSig);
+
+            // End of OpenSAML 2.6.4 logic
+            // -----------------------------------------------------------------------------
+
+        } catch (ValidationException ex) {
+            String logMsg = "Signature do not confirm to SAML signature profile. Possible XML Signature " +
+                    "Wrapping  Attack!";
+            AUDIT_LOG.warn(logMsg);
+            if (log.isDebugEnabled()) {
+                log.debug(logMsg, ex);
+            }
+            throw new SSOAgentException(logMsg, ex);
+        }
+
```

Comment:
Fixes coming from opensaml 2.6.4 version is not needed here. This component imports opensaml in 2.6.0 to 3.0.0 version range which already has these fixes